Add support for Windows memory-pressure eviction

marosset opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation


Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR enabled memory-pressure eviction for the `memory.available` signal to work on Windows and sets the default hard eviction threshold to `500Mi`

On Windows we'll use the GlobalMemoryStatusEx windows API call to get the available and total physical memory and use that to determine if the node is experiencing memory pressure. 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of #119184 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a missing behavior where Windows nodes did not implement memory-pressure eviction.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

marosset committed:
Add support for memory pressure evictiong on Windows

Signed-off-by: Mark Rossetti <marosset@microsoft.com>

marosset committed:
Adding e2e test to validate memory-pressure eviction on Windows

Signed-off-by: Mark Rossetti <marosset@microsoft.com>

marosset commented:
/test ls

marosset commented:
/sig node windows
/area kubelet
/triage accepted
/milestone v1.30

jsturtevant commented:
/priority important-soon

kannon92 commented on pkg/kubelet/eviction/defaults_windows.go:
```diff
@@ -0,0 +1,26 @@
+// DefaultEvictionHard includes default options for hard eviction.
+var DefaultEvictionHard = map[string]string{
+	"memory.available": "500Mi",
+	"nodefs.available": "10%",
```
Can kubernetes in windows support a separate disk? If so, we should probably add imagefs.available as a default option?

I guess inodes are linux only so only that should be needed.

marosset replied:
I need to check if Windows supports separate disks or not - I don't remember :-/

marosset replied:
ah - good to know!

marosset replied:
It does look like the kubelet flag `--root-dir` works on windows and can target a directory on another drive.
Let me add the `image.fs.available` default to the windows defaults.
Thanks @kannon92

kannon92 replied:
So its not a kubelet setting actually. The container runtime is allowed to say I want my images/containers stored on a separate disk from kubelet. So I'd look into windows containers (I admit I know nothing about this area so I don't even know what that is).

I'm actually working on a blog post around this area: https://github.com/kubernetes/website/pull/43971

The configuration for the separate filesystem comes from storage configurations for crio/containerd actually. Kubelet doesn't know or care about that.

marosset replied:
got it. either way i added the default back in for windows :)

marosset commented:
/test ls

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow

kannon92 commented on test/e2e/windows/eviction.go:
```diff
@@ -0,0 +1,170 @@
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+var _ = sigDescribe(feature.Windows, "Eviction", framework.WithSerial(), framework.WithSlow(), (func() {
```
To match other eviction tests, should this be considered disruptive also?

marosset replied:
I'll add that

kannon92 commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -847,10 +849,33 @@ func makeSignalObservations(summary *statsapi.Summary) (signalObservations, stat
-			available: resource.NewQuantity(int64(*memory.AvailableBytes), resource.BinarySI),
-			capacity:  resource.NewQuantity(int64(*memory.AvailableBytes+*memory.WorkingSetBytes), resource.BinarySI),
-			time:      memory.Time,
+		if runtime.GOOS == "windows" {
```
Is it possible to add a unit test to cover this code? We have a lot of tests for helpers.

marosset replied:
I updated the existing test to change behavior when running on Windows.
This approach seemed less disruptive to the existing code and should give adequate coverage.
We do run the unit tests on Windows nodes regularly and monitor the updates - https://testgrid.k8s.io/sig-windows-signal#windows-unit-master

kannon92 commented on test/e2e/windows/eviction.go:
```diff
@@ -0,0 +1,170 @@
+			e2eskipper.Skipf("No Windows nodes with hard memory-pressure eviction found")
+		}
+
+		ginkgo.By("Scheduling a pod that requests and consumse 500Mi of Memory")
```
```suggestion
		ginkgo.By("Scheduling a pod that requests and consumes 500Mi of Memory")
```

kannon92 commented on test/e2e/windows/eviction.go:
```diff
@@ -0,0 +1,170 @@
+
+		ginkgo.By("Waiting for node.kubernetes.io/memory-pressure taint to be removed")
+		// ensure e2e test framework catches the memory-pressure taint
+		time.Sleep(20 * time.Second)
```
Could you move this to a constant at the top of this file? And name it. If we hit any flakes it would be nice to be able to bump this easily.

kannon92 commented on test/e2e/windows/memory_limits.go:
```diff
@@ -217,18 +202,33 @@ func getNodeMemory(ctx context.Context, f *framework.Framework) nodeMemory {
 	return nodeMem
 }
 
+// getNodeMemory populates a nodeMemory struct with information from the first
```
I don't follow what you mean by "first".

kannon92 replied:
Ah I see. You just mean the first element of the list of windows nodes.

marosset replied:
Let me update that :)

kannon92 commented on test/e2e/windows/eviction.go:
```diff
@@ -0,0 +1,170 @@
+		}
+
+		ginkgo.By("Scheduling a pod that requests and consumse 500Mi of Memory")
+
```
Few thoughts on this test case.

Are we protecting ourselves if we ever increase memory for the node size? The hard coded 500Mi?

Is it possible to read the amount of the memory of the node and calculate the MB for the pod based on that?

I don't understand why scheduling two pods of 500 Mi would cause memory pressure here? Do you need to set evictionHard limits?

marosset replied:
The first pod in the test will consume 500Mi of memory, the second pod will consume the rest of the node's memory in 100Mi chunks using https://learn.microsoft.com/en-us/sysinternals/downloads/testlimit which is in the Windows `resource-consumer` test image.

Let me try and get the test passing as is in CI (I was using a much smaller node when testing locally) then I'll update the logic to be smarter here.

marosset replied:
I updated this a little and added some more comments - please take a look!

sftim commented:
/retitle Add support for Windows memory-pressure eviction

Changelog suggestion
```diff
-Memory-pressure eviction now works for Windows nodes!
+Fixed a missing behavior where Windows nodes did not implement memory-pressure eviction.
```

marosset commented:
> /retitle Add support for Windows memory-pressure eviction
> 
> Changelog suggestion
> 
> ```diff
> -Memory-pressure eviction now works for Windows nodes!
> +Fixed a missing behavior where Windows nodes did not implement memory-pressure eviction.
> ```

Updated - thanks!

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow

kannon92 commented:
/assign @kannon92

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow

marosset commented:
The e2e test I added is failing because the eviction manager is evicting some imagepuller pods left on the nodes
The pod I'd like the to get evicted during the test was ranked as second.
I'm going to play with the test some to get it to stable

```
I0201 23:18:29.241621    4304 eviction_manager.go:368] "Eviction manager: attempting to reclaim" resourceName="memory"
I0201 23:18:29.241621    4304 eviction_manager.go:379] "Eviction manager: must evict pod(s) to reclaim" resourceName="memory"
I0201 23:18:29.241621    4304 eviction_manager.go:397] "Eviction manager: pods ranked for eviction" pods=["img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm","eviction-test-windows-1069/pod2","calico-system/calico-node-windows-cw55g","kube-system/kube-proxy-windows-68lmk","img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-httpd-2.4.38-4-lxpkr","img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-agnhost-2.45-prw59","kube-system/containerd-logger-2kb79","img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-nginx-1.14-4-8sbth","eviction-test-windows-1069/pod1","kube-system/csi-proxy-rtb8p","kube-system/cloud-node-manager-windows-r6pwp"]
I0201 23:18:29.725263    4304 kubelet_node_status.go:679] "Recording event message for node" node="capz-conf-fmslq" event="NodeHasInsufficientMemory"
I0201 23:18:29.731430    4304 kubelet.go:2435] "SyncLoop (PLEG): event for pod" pod="img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm" event={"ID":"1dff04c7-9a55-4f36-91b2-14177bda5f7a","Type":"ContainerDied","Data":"0f26216dfa830fed61b1546666ebbb1807fc498f6efd517583b19e2a748838ec"}
I0201 23:18:29.731565    4304 fake_topology_manager.go:76] "RemoveContainer" containerID="a5308f218faecc60b3d73d3270df8b7fc3cc9c1dacf316ad1ebbaa16f7e53c94"
I0201 23:18:29.943251    4304 kubelet.go:2410] "SyncLoop UPDATE" source="api" pods=["img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm"]
I0201 23:18:32.958508    4304 eviction_manager.go:185] "eviction manager: hard memory eviction threshold crossed"
I0201 23:18:33.087495    4304 helpers.go:968] "Eviction manager: no observation found for eviction signal" signal="containerfs.inodesFree"
I0201 23:18:33.087495    4304 helpers.go:968] "Eviction manager: no observation found for eviction signal" signal="containerfs.inodesFree"
I0201 23:18:33.250073    4304 util.go:48] "No ready sandbox for pod can be found. Need to start a new one" pod="img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm"
I0201 23:18:33.250073    4304 eviction_manager.go:618] "Eviction manager: pod is evicted successfully" pod="img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm"
I0201 23:18:33.250749    4304 eviction_manager.go:207] "Eviction manager: pods evicted, waiting for pod to be cleaned up" pods=["img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm"]
```

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow

marosset commented:
> The e2e test I added is failing because the eviction manager is evicting some imagepuller pods left on the nodes The pod I'd like the to get evicted during the test was ranked as second. I'm going to play with the test some to get it to stable
> 
> ```
> I0201 23:18:29.241621    4304 eviction_manager.go:368] "Eviction manager: attempting to reclaim" resourceName="memory"
> I0201 23:18:29.241621    4304 eviction_manager.go:379] "Eviction manager: must evict pod(s) to reclaim" resourceName="memory"
> I0201 23:18:29.241621    4304 eviction_manager.go:397] "Eviction manager: pods ranked for eviction" pods=["img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm","eviction-test-windows-1069/pod2","calico-system/calico-node-windows-cw55g","kube-system/kube-proxy-windows-68lmk","img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-httpd-2.4.38-4-lxpkr","img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-agnhost-2.45-prw59","kube-system/containerd-logger-2kb79","img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-nginx-1.14-4-8sbth","eviction-test-windows-1069/pod1","kube-system/csi-proxy-rtb8p","kube-system/cloud-node-manager-windows-r6pwp"]
> I0201 23:18:29.725263    4304 kubelet_node_status.go:679] "Recording event message for node" node="capz-conf-fmslq" event="NodeHasInsufficientMemory"
> I0201 23:18:29.731430    4304 kubelet.go:2435] "SyncLoop (PLEG): event for pod" pod="img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm" event={"ID":"1dff04c7-9a55-4f36-91b2-14177bda5f7a","Type":"ContainerDied","Data":"0f26216dfa830fed61b1546666ebbb1807fc498f6efd517583b19e2a748838ec"}
> I0201 23:18:29.731565    4304 fake_topology_manager.go:76] "RemoveContainer" containerID="a5308f218faecc60b3d73d3270df8b7fc3cc9c1dacf316ad1ebbaa16f7e53c94"
> I0201 23:18:29.943251    4304 kubelet.go:2410] "SyncLoop UPDATE" source="api" pods=["img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm"]
> I0201 23:18:32.958508    4304 eviction_manager.go:185] "eviction manager: hard memory eviction threshold crossed"
> I0201 23:18:33.087495    4304 helpers.go:968] "Eviction manager: no observation found for eviction signal" signal="containerfs.inodesFree"
> I0201 23:18:33.087495    4304 helpers.go:968] "Eviction manager: no observation found for eviction signal" signal="containerfs.inodesFree"
> I0201 23:18:33.250073    4304 util.go:48] "No ready sandbox for pod can be found. Need to start a new one" pod="img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm"
> I0201 23:18:33.250073    4304 eviction_manager.go:618] "Eviction manager: pod is evicted successfully" pod="img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm"
> I0201 23:18:33.250749    4304 eviction_manager.go:207] "Eviction manager: pods evicted, waiting for pod to be cleaned up" pods=["img-puller-7559/img-pull-registry.k8s.io-e2e-test-images-busybox-1.36.1-1-lsbbm"]
> ```

I added some logic to delete the namespace all the image puller pods on the node we are running the eviction test on.
This should be OK since the test is marked as disruptive and can save ~10 mins from the test

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow

MaryamTavakkoli commented:
Hello! 
Release Signal shadow here.
The code freeze is starting [02:00 UTC Wednesday 6th March 2024 / 18:00 PDT Tuesday 5th March 2024](https://everytimezone.com/s/5056bace) (less than a month from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As this PR is tagged for 1.30, is it still planned for this release?

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow

marosset commented:
pull-kubernetes-e2e-capz-windows-serial-slow

marosset commented:
/test pull-kubernetes-e2e-capz-windows-master

marosset commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow
Since https://github.com/kubernetes/test-infra/pull/31989 updated the job timeout

marosset commented:
/test pull-kubernetes-e2e-capz-windows-serial-slow

marosset commented:
/test pull-kubernetes-e2e-capz-windows-master
/test pull-kubernetes-e2e-capz-windows-serial-slow

marosset commented:
The serial-slow test pass is hitting other flakes :(
/test pull-kubernetes-e2e-capz-windows-serial-slow

jsturtevant commented on pkg/kubelet/eviction/memory_threshold_notifier_windows.go:
```diff
@@ -0,0 +1,96 @@
+
+			// Get global memory usage on Windows
+			availablePhysMem, totalPhysMem, err := winstats.GetAvailableAndTotalPhysicalMemory()
+			//mem, err := GlobalMemoryStatusEx()
```
```suggestion

```

jsturtevant commented on pkg/kubelet/eviction/helpers_test.go:
```diff
@@ -1319,11 +1322,25 @@ func TestMakeSignalObservations(t *testing.T) {
+		if expectedBytes := int64(totalPhysMem); memQuantity.capacity.Value() != expectedBytes {
+			t.Errorf("Expected %v, actual: %v", expectedBytes, memQuantity.capacity.Value())
+		}
+		if math.Abs(float64(int64(availablePhysMem)-memQuantity.available.Value())) > (10 * 1024 * 1024) {
+			t.Errorf("Reported available memory is not within 10Mb of expected value. Expected: %v, actual: %v", availablePhysMem, memQuantity.available.Value())
+		}
```
This is to handle fluctuation in the memory since it is a queried value?  Do we need to take into account the other direction (where memory might have been freed)?

marosset replied:
This checks the absolute value of the delta so i think that should cover if more memory has been freed too?

jsturtevant replied:
thanks I missed that!

jsturtevant commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -847,10 +849,30 @@ func makeSignalObservations(summary *statsapi.Summary) (signalObservations, stat
+		if runtime.GOOS == "windows" {
+			// Don't use stats summary to compute available memory on Windows because it does not take into account
+			// system process memory usage (which can be significant). Instead use win32 api's to query global
+			// memory usage.
```
Is this an issue with the Stats summary and related to https://github.com/kubernetes/kubernetes/issues/112111?  Should we be updating the stats summary to use the correct value here instead of querying it directly?

marosset replied:
It has been a few weeks since I last looked at this but I don't think issue was related to Stats summary being incorrect - it just works different from linux. On Windows OS overhead is a lot compared to Linux and it made more sense to just use the perf counters. I'll try and remember exactly why...

jsturtevant replied:
I am wondering what the node stats are used for in Windows, I thought this would be one of the main use cases.  The reason for asking is I think it would clean up this nested if statements here as well as the check in the tests for being with in a range.

marosset replied:
I'm going to look into this again

marosset replied:
I finally got a chance to take a deep look at this and I'm still thinking that using total and available physical memory is the right thing to do for eviction thresholds.

Eviction thresholds can be specified as a percentage of total memory (< 5%) or a value (< 500Mb). Windows always uses virtual memory and trying to determine these values based on what we have in the stats summary / available in perf counters is too unpredictable.

I agree that there can be issues in how memory usage is reported in /stats/summary but don't see an easy way to fix that that would also allow eviction to work predictably.

I wrote a small powershell script to dump memory info (both total / available physical memory and variety of memory / process memory perf counters)

https://gist.githubusercontent.com/marosset/8334544e261d57e756f13514a3202a62/raw/e1b0b5161c18514b3bb6b335de1a2660effef901/memory.ps1

When running on different machines with different amounts of memory here are some example outputs

```console
 .\memory.ps1
Stats from systeminfo / GlobalMemoryStatusEx
  Total Physical Memory Mb:  4095.55078125
  Free Physical Memory Mb :  1903.83984375
  Percent Usage           :  53.5144368746191

Status from perf counters
Available Bytes MB               : 1885.375
Committed Bytes MB               : 2426.89453125
Cache Bytes MB                   : 87.328125
Working Set (total) MB           : 2374.515625
Working Set - Private (total) MB : 992.08203125
Private (total) MB               : 1672.94140625

Available Bytes + committed Bytes MB         :  4312.26953125
Available Bytes + Working Set MB             :  4259.890625
Available Bytes + Working Set - Cached MB    :  4172.5625
Available Bytes + Private Bytes MB           :  3558.31640625
Available Bytes + (Working Set - Private) MB :  2877.45703125
```
 
and 

```console
.\memory.ps1
Stats from systeminfo / GlobalMemoryStatusEx
  Total Physical Memory Mb:  65401.28125
  Free Physical Memory Mb :  24365.81640625
  Percent Usage           :  62.744129869397625

Status from perf counters
Available Bytes MB               : 24381.48046875
Committed Bytes MB               : 55918.75390625
Cache Bytes MB                   : 920.1328125
Working Set (total) MB           : 44125.765625
Working Set - Private (total) MB : 31531.4765625
Private (total) MB               : 48290.07421875

Available Bytes + committed Bytes MB         :  80300.234375
Available Bytes + Working Set MB             :  68507.24609375
Available Bytes + Working Set - Cached MB    :  67587.11328125
Available Bytes + Private Bytes MB           :  72671.5546875
Available Bytes + (Working Set - Private) MB :  55912.95703125
```

I couldn't find a combination of metrics reported from perf metrics that reasonably estimate memory usage as close as what this PR currently does.

@jsturtevant - let's discuss in the next SIG-Windows meeting if you'll be able to attend.

jsturtevant commented on test/e2e/windows/eviction.go:
```diff
@@ -0,0 +1,199 @@
+)
+
+const (
+	// It can take 10-15 seconds for node memory-pressure taint to show up on the node
+	// so we'll wait 20 seconds for the taint to show up so the e2e test case can catch
```
should we add some more buffer to avoid flakes due to other delays?

marosset replied:
I can do that

marosset replied:
Updated the delay to 45 seconds

jsturtevant commented on test/e2e/windows/eviction.go:
```diff
@@ -0,0 +1,199 @@
+const (
+	// It can take 10-15 seconds for node memory-pressure taint to show up on the node
+	// so we'll wait 20 seconds for the taint to show up so the e2e test case can catch
+	// it and the wait for the taint to be removed so other serial/slow tests can run
+	// against the same node.
```
what happens if this test fails and doesn't remove the this taint? Can we add a deferred removal in the case where the tests fails, where it checks for the existence of and if still there removes it?

marosset replied:
The test doesn't remove the taint, the taint gets removed by the kubelet / eviction manager after the node is no longer experiencing memory / cpu pressure.
In this case I think we shouldn't manually remove the taint. If the test fails then the namespace created for the test will get deleted - which should delete the pods - and then everything should go back to normal after a few minutes.
The e2e test pass also will pause until all the nodes are schedulable again.

Let me know what you think

jsturtevant replied:
This makes sense, thanks for explaining

jsturtevant commented on test/e2e/windows/memory_limits.go:
```diff
@@ -217,18 +202,34 @@ func getNodeMemory(ctx context.Context, f *framework.Framework) nodeMemory {
+	})
+	framework.ExpectNoError(err)
+
+	// Assuming that agent nodes have the same config
+	// Make sure there is >0 agent nodes, then use the first one for info
+	gomega.Expect(nodeList.Items).ToNot(gomega.BeEmpty())
```
We've had other tests fail becuase the nodes had different config (https://github.com/kubernetes/kubernetes/issues/106608).  This looks like it just moved logic but maybe need to consider opening an issue and fixing like we did for the other set of tests.

MaryamTavakkoli commented:
Hello!
Release Signal shadow here.
Just a reminder that the code freeze is starting [02:00 UTC Wednesday 6th March 2024 / 18:00 PDT Tuesday 5th March 2024](https://everytimezone.com/s/5056bace) in a week, and we want to ensure that each PR has a chance to be merged on time.

liggitt commented:
/milestone clear

marosset commented:
/test pull-kubernetes-e2e-capz-windows-master
cluster didn't come up

marosset commented:
@kannon92 @jsturtevant - Can you take another look at this when you get a chance?
The 2 main updates since your last reviews are:
- The eviction signal was switched to be based on the system the available / used pages of committed memory (managed by windows) instead of the available / used amount of physical memory
- Updated the stats/summary endpoint for Windows to report available / used commit memory so eviction manager could avoid querying this info from Windows APIs.
PTAL and thanks!

marosset commented:
/milestone v1.31

jsturtevant commented on pkg/kubelet/winstats/perfcounter_nodestats_unsupported.go:
```diff
@@ -0,0 +1,26 @@
+
+import "errors"
+
+func GetAvailableAndTotalPhysicalMemory() (uint64, uint64, error) {
```
This function doesn't look to be used outside windows, I don't think this is required anymore (was maybe part of previous versions of the PR?)

marosset replied:
you are correct, let me remove it

marosset replied:
updated!

kannon92 commented:
/unassign

I’m on leave so maybe @rphillips, @haircommander or @harche could help with a review here.

harche commented:
> I’m on leave so maybe @rphillips, @haircommander or @harche could help with a review here.

I am not sure if any of us have expertise in windows to review this patch.

marosset commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -845,12 +846,31 @@ func makeSignalObservations(summary *statsapi.Summary) (signalObservations, stat
 	statsFunc := cachedStatsFunc(summary.Pods)
 	// build an evaluation context for current eviction signals
 	result := signalObservations{}
+	if runtime.GOOS == "windows" {
```
updates

marosset commented:
> > I’m on leave so maybe @rphillips, @haircommander or @harche could help with a review here.
> 
> I am not sure if any of us have expertise in windows to review this patch.

@harche - would you be able to review some of the non-windows specific kubelet changes?

marosset commented:
/test pull-kubernetes-integration

jsturtevant commented:
looks like helm failed to install calico
/retest

jsturtevant commented:
looks like the integration test is passing the tests but timing out. Looks like a few fixes went in recently, @marosset could you rebase to pick those up?

Will run one more time just incase:
/retest

jsturtevant commented on pkg/kubelet/eviction/defaults_windows.go:
```diff
@@ -0,0 +1,27 @@
+// DefaultEvictionHard includes default options for hard eviction.
+var DefaultEvictionHard = map[string]string{
+	"memory.available":  "500Mi",
+	"nodefs.available":  "10%",
+	"imagefs.available": "15%",
```
Do we need or want these?  Will these affect how storagge evicts?

marosset replied:
The nodefs and imagefs are already the defaults for Windows.
Initially I had left them out when I made new Windows defaults but someone said I should leave them in (let me try and find the discussion)

marosset replied:
https://github.com/kubernetes/kubernetes/pull/122922#discussion_r1462516048

haircommander commented on pkg/kubelet/eviction/memory_threshold_notifier_windows.go:
```diff
@@ -0,0 +1,98 @@
+func (m *windowsMemoryThresholdNotifier) Start() {
+	klog.InfoS("Eviction manager: starting windowsMemoryThresholdNotifier", "notifier", m.Description())
+	go func() {
+		for true {
```
personally, this seems involved enough and doesn't use too many scoped variables to warrant a separate function

marosset replied:
updated - PTAL

haircommander commented:
linux pieces LGTM, I left one note on the windows side though impl wise it's also not in my area of expertise

marosset commented on pkg/kubelet/eviction/memory_threshold_notifier_windows.go:
```diff
@@ -0,0 +1,98 @@
+	}
+}
+
+func (m *windowsMemoryThresholdNotifier) UpdateThreshold(summary *statsapi.Summary) error {
```
That stats summary is actually generated and passed in by the eviction manager.
We do need to keep this for two reasons
- This is part of the `ThresholdNotifier` interface and
- The memory threshold notifier is responsible for raising the signal a node is experiencing memory pressure but the eviction manager also uses the stats summary to determine when a node is longer experience memory pressure so both need to be able to consume the win32 api call.
 
There shouldn't be much cpu load caused by this call and we call it relatively infrequently

jsturtevant commented:
/lgtm

marosset commented:
/test pull-kubernetes-e2e-capz-windows-master

marosset commented:
@mrunalp @dchen1107 - Can one of you take a look at this for node approval?

marosset commented:
I'm investigating why the windows presubmit jobs are failing at cluster creation too

marosset commented:
/test pull-kubernetes-e2e-capz-windows-master

SergeyKanzhelev commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -187,7 +188,8 @@ func (m *managerImpl) Start(diskInfoProvider DiskInfoProvider, podFunc ActivePod
 	}
-	if m.config.KernelMemcgNotification {
+	klog.InfoS("Eviction manager: starting control loop")
+	if m.config.KernelMemcgNotification || runtime.GOOS == "windows" {
```
what was the behavior before when people tried to set `KernelMemcgNotification` to true? I would think it will fail kubelet?

marosset commented:
/test pull-kubernetes-e2e-capz-windows-master

marosset commented:
/test pull-kubernetes-e2e-capz-windows-master
There seems to be some capacity issues in azure recently :-/

rphillips commented on pkg/kubelet/eviction/helpers_others.go:
```diff
@@ -0,0 +1,37 @@
+//go:build !windows
```
This file can safely be called helpers_unix.go or helpers_linux.go

marosset replied:
We do build for `darwin` now which I believe should have the same implementation as linux.
I think _others and _windows is cleaner but let me know and I can make a _linux, _windows, _darwin (or _others)

rphillips replied:
This is probably ok for now. Thanks!

rphillips commented on pkg/kubelet/eviction/memory_threshold_notifier_others.go:
```diff
@@ -0,0 +1,123 @@
+//go:build !windows
```
same comment... probably _linux or _unix.go.

rphillips commented on pkg/kubelet/eviction/memory_threshold_notifier_windows.go:
```diff
@@ -0,0 +1,101 @@
+func (m *windowsMemoryThresholdNotifier) Description() string {
+	var hard, allocatable string
+	if isHardEvictionThreshold(m.threshold) {
+		hard = "hard "
```
nit: don't put the spaces in these assignments, and place the spaces within the sprintf below.

rphillips commented:
Overall, lgtm.

marosset commented:
/assign @yujuhong 
PTAL - thanks!

marosset commented:
/test pull-kubernetes-node-e2e-containerd

marosset commented:
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-capz-windows-master

Vyom-Yadav commented:
Hello @marosset,

I see this PR is being actively worked on, just a reminder about code freeze ⬇️ 

Reminder: Code freeze starts `02:00 UTC Wednesday 24th July 2024 / 19:00 PDT Tuesday 23rd July 2024` (about 1 week from now). Please make sure the PR has both `lgtm` and `approved` labels before the code freeze.

marosset commented:
/assign @mrunalp

rphillips commented:
/lgtm

mrunalp commented on pkg/kubelet/eviction/helpers_test.go:
```diff
@@ -1289,7 +1289,7 @@ func TestMakeSignalObservations(t *testing.T) {
 				// Used for memory signal observations on windows
 				{
-					Name: statsapi.SystemContainerPhysicalMemory,
+					Name: statsapi.SystemContainerWindowsGlobalCommitMemory,
```
nit: Can we make the commits additive?

marosset replied:
I'll rebase into a fewer commits.
I had let some of the iterative commits in to facilitate discussion in sig-windows but now that we have reached agreement this can and should go in as a single commit.

mrunalp commented on pkg/kubelet/eviction/memory_threshold_notifier_windows.go:
```diff
@@ -90,12 +90,12 @@ func (m *windowsMemoryThresholdNotifier) UpdateThreshold(summary *statsapi.Summa
 func (m *windowsMemoryThresholdNotifier) Description() string {
 	var hard, allocatable string
 	if isHardEvictionThreshold(m.threshold) {
-		hard = "hard "
```
nit: Can we move these changes to the commit that first introduced them earlier?

marosset replied:
I'll rebase the changes down to 1 or 2 commits.

mrunalp commented on pkg/kubelet/eviction/defaults_windows.go:
```diff
@@ -0,0 +1,27 @@
+
+// DefaultEvictionHard includes default options for hard eviction.
+var DefaultEvictionHard = map[string]string{
+	"memory.available":  "500Mi",
```
On Linux, we have this at `100Mi`. Should we document the divergence here?

marosset replied:
I'm going to update the k8s.io website with details about how memory pressure eviction will work on Windows.
Would it be sufficient to mention it there?

mrunalp replied:
Sounds good!

marosset replied:
updated in the code too

marosset commented:
The windows job will likely fail because of https://github.com/kubernetes/kubernetes/issues/126180 
We are working to resolve that but it shouldn't have any impact on this PR

marosset commented:
/skip pull-kubernetes-e2e-capz-windows-master
There are some know / unrelated failures mentioned above

marosset commented:
/skip pull-kubernetes-e2e-capz-windows-master
